Extensible Userspace Proxy

DirectXMan12 opened the pull request:
This PR refactors the userspace proxy to allow for custom proxy socket implementations.
It changes the the ProxySocket interface to ensure that other packages can properly implement it (making sure all arguments are publicly exposed types, etc), and adds in a mechanism for an implementation to create an instance of the userspace proxy with a non-standard ProxySocket.
Custom ProxySockets are useful to inject additional logic into the actual proxying.  For example, our idling proxier uses a custom proxy socket to hold connections and notify the cluster that idled scalable resources need to be woken up.

Also-Authored-By: Ben Bennett bbennett@redhat.com

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29465)

<!-- Reviewable:end -->

knobunc committed:
Userspace proxy should remove conntrack entries

This changes the userspace proxy so that it cleans up its conntrack
settings when a service is removed (as the iptables proxy already
does).  This could theoretically cause problems when a UDP service
as deleted and recreated quickly (with the same IP address).  As
long as packets from the same UDP source IP and port were going to
the same destination IP and port, the the conntrack would apply and
the packets would be sent to the old destination.

This is astronomically unlikely if you did not specify the IP address
to use in the service, and even then, only happens with an "established"
UDP connection.  However, in cases where a service could be "switched"
between using the iptables proxy and the userspace proxy, this case
becomes much more frequent.

DirectXMan12 commented:
(The second commit here is a fix for a bug in the userspace proxy that we discovered while implementing the unidling proxy.  It's easiest to see when you have a UDP port on service switching between a userspace-based proxy and the iptables proxy, but could hypothetically occur in normal userspace proxy usage.  I can split that off into another PR, if desired.)

DirectXMan12 commented:
cc @smarterclayton @knobunc @ncdc

ncdc commented:
cc @kubernetes/sig-network @kubernetes/rh-networking

ncdc commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -139,7 +148,15 @@ func IsProxyLocked(err error) bool {
+}
+
+// NewCustomProxier functions similarly to NewProxier, returing a new Proxier
+// for the given LadBalancer and address.  The new proxier is constructed using
```
LoadBalacner

DirectXMan12 replied:
LoadBalancer 😉

ncdc commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -1049,3 +1090,32 @@ func isClosedError(err error) bool {
+	for _, ip := range svcIPs {
+		glog.V(2).Infof("Deleting connection tracking state for service IP %s", ip)
+		err := proxier.execConntrackTool("-D", "--orig-dst", ip, "-p", "udp")
+		if err != nil && !strings.Contains(err.Error(), noConnectionToDelete) {
```
How brittle is the string comparison?

DirectXMan12 replied:
I believe this is pulled from the iptables code that we're already using.
@knobunc can probably give a better answer.

knobunc replied:
Yes, this was a direct port from pkg/proxy/iptables/proxier.go.  I forgot that we moved it over, so now we are changing kubernetes we could make the function common and just call it from both places.

DirectXMan12 replied:
Done :check:

ncdc commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -1049,3 +1090,32 @@ func isClosedError(err error) bool {
+func (proxier *Proxier) execConntrackTool(parameters ...string) error {
+	conntrackPath, err := proxier.exec.LookPath("conntrack")
+	if err != nil {
+		return fmt.Errorf("Error looking for path of conntrack: %v", err)
```
error messages start with a lowercase letter

ncdc commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -1049,3 +1090,32 @@ func isClosedError(err error) bool {
+	}
+	output, err := proxier.exec.Command(conntrackPath, parameters...).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("Conntrack command returned: %q, error message: %s", string(output), err)
```
lowercase conntrack

DirectXMan12 commented:
(that looks like it might be a flake -- I see a message about an issue pulling the images from gcr.io)

ncdc commented:
@k8s-bot e2e test this issue: #26135

ncdc commented:
@DirectXMan12 unit test compilation is also failing

FAIL	k8s.io/kubernetes/pkg/proxy/userspace [build failed]

DirectXMan12 commented:
@ncdc thanks.  Those tests results are easy to misread.

DirectXMan12 commented:
I've moved the conntrack methods used by both proxies into `pkg/proxy/util`.

thockin commented:
This is a very high-risk change.  The userpsace code is more delicate than I am happy about.

a) can this be broken down into smaller, easier to verify steps?

b) can this be a new proxymode rather than changing userspace?

DirectXMan12 commented:
So, this is mostly just a case of making a few methods and fields in the userspace proxy public instead of private.  There's not really any mechanical change in the first commit.

The UDP conntrack change can be fairly easily split into a separate PR (it's already a separate commit), since that's what I'd imagine would be the most "dangerous", since it's actually a mechanical change.

thockin commented:
I am TERRIFIED of changes to that userspace code.  If you can break this
into more digestible pieces, I sure would appreicate it.

On Wed, Aug 3, 2016 at 7:41 AM, Solly Ross <notifications@github.com> wrote:

> So, this is mostly just a case of making a few methods and fields in the
> userspace proxy public instead of private. There's not really any
> mechanical change in the first commit.
>
> The UDP conntrack change can be fairly easily split into a separate PR
> (it's already a separate commit), since that's what I'd imagine would be
> the most "dangerous", since it's actually a mechanical change.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29465#issuecomment-237245194>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVAgPH9OOVn6ZyzDRrwYd-fyIJff_ks5qcKiJgaJpZM4JTCFd>
> .
>

DirectXMan12 commented:
So, the first commit is mostly just renames.  The new code is a) allowing a custom proxy-port creation method to be passed, b) adding the serviceRef, and c) adding a new method to the load balancer to check if a service has endpoints.  Do you want me to put those pieces into a separate commit?  The rest of the first commit is renames (moving from private to public)?

Alternatively, did you just want the UDP changes moved from being in a second commit to a separate PR?

DirectXMan12 committed:
Userspace Proxy: allow check for endpoints on svc

This commit adds a method to the `LoadBalancer` interface in the
userspace proxy which allows consumers of the `LoadBalancer` to check if
it thinks a given service has endpoints available.

DirectXMan12 committed:
Userspace Proxy: Make ProxySocket Implementable

This commit makes it possible for the `ProxySocket` interface to be
implemented by types outside of the `userspace` package.  It mainly just
exposes relevant types and fields as public.

DirectXMan12 committed:
Userspace Proxy: Allow any ProxySocket in Proxier

This commit adds a new method for constructing userspace proxiers,
`NewCustomProxier`.  `NewCustomProxier` functions identically to
`NewProxier`, except that it allows a custom constructor method to
be passed in to construct instances of ProxySocket.

DirectXMan12 committed:
Userspace Proxy: Expose ProxySocket utility funcs

This commit exposes several utility functions that are valuable for
implementing custom ProxySockets.

DirectXMan12 committed:
Userspace Proxy: Keep ref to service being proxied

This commit makes the userspace proxy keep an ObjectReference to the
service being proxied.  This allows the consumers of the `ServiceInfo`
struct, like `ProxySockets` to emit events about or otherwise refer to
the service.

DirectXMan12 commented:
Alright, I've tried to break the PR up a bit more into still functioning chunks.  Hopefully, that's a bit easier to review.

tmrts commented on pkg/proxy/userspace/roundrobin.go:
```diff
@@ -120,6 +120,14 @@ func isSessionAffinity(affinity *affinityPolicy) bool {
+	lb.lock.Lock()
+	defer lb.lock.Unlock()
+	state, exists := lb.services[svcPort]
+	return exists && state != nil && len(state.endpoints) > 0
```
why would the state be `nil` if `exists` is true?

DirectXMan12 replied:
some of the other code which uses `lb.services` explicitly deals with the `state == nil` check separately from the `exists` check (see `OnEndpointsUpdate`), while other code does not.  It looks like nothing currently ever assigns `nil` to an `lb.services` entry (although it could), so it's *probably* ok to remove the `state != nil` case.

tmrts replied:
@DirectXMan12 can you refactor it or add a TODO item? Because, the condition will only get worse as other people will do the same and perform `nil` checks since the current code does it

tmrts commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -42,14 +42,14 @@ type portal struct {
-type serviceInfo struct {
+type ServiceInfo struct {
+	Timeout       time.Duration
+	ActiveClients *ClientCache
```
nit: missing exported type comment mentioning that `ActiveClients` is a cache and it might be stale

DirectXMan12 commented:
@thockin can you take another look if possible, now that this is broken up into smaller chunks?

danwinship reviewed (commented):
The parts I understand basically LGTM. (I can't really comment on the ObjectReference commit.)

There's a typo in the commit message on the conntrack commit: "theoretrically". Although I'm not sure I agree that the "problem" it describes is a "problem" anyway; if you delete the Service and recreate it, I think it's totally reasonable for existing connections to get dropped.

knobunc commented:
@danwinship it's a problem because the "connections" are UDP "connections" and the source ip:port -> service ip:port match and it will send new packets to the old backend.  TCP did not exhibit this behavior because the source port changes.

danwinship commented:
> and it will send new packets to the old backend

ah, I missed that: I thought the problem was "without this patch the service may lose packets", not "without this patch the service may get packets it shouldn't".

DirectXMan12 commented:
Ok, fixed the typo in the commit message and your comments.  Can you (or anyone else in @kubernetes/sig-network) give a final review?

DirectXMan12 commented:
@k8s-bot gci gke e2e test this

flaked on #33388

DirectXMan12 commented:
As for the unit tests, they appear to have failed in the ReplicaSet tests... I didn't change anything there, so it might be a flake?  It's strange looking.

smarterclayton commented:
Can you rebase so we can merge this (unless Tim wants to stop it now)?

DirectXMan12 commented:
@smarterclayton rebased

DirectXMan12 commented:
@k8s-bot verify test this

DirectXMan12 commented:
@k8s-bot cvm gce e2e test this

flake ##36379

DirectXMan12 commented:
ping @thockin can I get a review/LGTM on this?

DirectXMan12 commented:
@dcbw a review would be appreciated as well

dcbw commented:
@DirectXMan12 @thockin @knobunc on the surface, the PR LGTM.  Almost all of the commits are simply moving code around or changing the names of types.  I'm happy to go through it with a finer-toothed comb, but from a 30 minute review I think it looks fine.

DirectXMan12 commented:
flake #39471

@k8s-bot unit test this

thockin commented:
Just a couple nits and then LGTM.

DirectXMan12 commented on pkg/proxy/userspace/proxysocket.go:
```diff
@@ -188,19 +188,19 @@ func (udp *udpProxySocket) Addr() net.Addr {
-	mu      sync.Mutex
-	clients map[string]net.Conn // addr string -> connection
+type ClientCache struct {
+	sync.Mutex
```
ack, changed. 

From a curiosity perspective, is there a particular reason?  `ClientCache.Lock()` looks nicer to me than `ClientCache.Mu.Lock()`, and IMO more clearly conveys the direct intent of the action (but that's really just a matter of opinion).

thockin replied:
I think it's just convention - it's documented at https://github.com/kubernetes/community/blob/master/contributors/devel/coding-conventions.md, which also says it should be called "lock".  My C++ tingles in exposing a lock, which is why I suggested method-wrapping.

DirectXMan12 commented on pkg/proxy/userspace/proxysocket.go:
```diff
@@ -188,19 +188,19 @@ func (udp *udpProxySocket) Addr() net.Addr {
-	clients map[string]net.Conn // addr string -> connection
+type ClientCache struct {
+	sync.Mutex
+	Clients map[string]net.Conn // addr string -> connection
```
eh, there are situations where "lock, get, set if missing" is a bit nicer with the ability to directly lock.

DirectXMan12 commented:
@thockin thanks!  Updated to fix the nits, should be good to go.

thockin commented:
I'm approving now, renaming Mu to Lock can be a followup if this merges first, and consider method wrapping as a cleaner interface, maybe with a TrySet() or something, if that is a real need.  YAGNI

DirectXMan12 commented:
> onsider method wrapping as a cleaner interface, maybe with a TrySet() or something, if that is a real need. YAGNI

Will do.  I actually meant there were a couple of places we *currently* do a check, if missing initialize, set kind of operation, and I need to think through a bit more to make sure I don't break any locking logic (since the current logic is the same as the current proxy code).

DirectXMan12 commented:
Flake #41889 

@k8s-bot gce etcd3 e2e test this

DirectXMan12 commented:
Does this need a milestone assigned to actually merge?

thockin commented:
FWIW this PR is a good example of such a refactoring - each commit was digestible and self-contained.  Thanks

ncdc commented:
@k8s-bot gce etcd3 e2e test this #41889 #42099

ncdc commented:
@k8s-bot bazel test this #39975
